staging/src/k8s.io/client-go/discovery/cached/memcache.go

```diff
@@ -19,51 +19,102 @@ package cached
 import (
 	"errors"
 	"fmt"
+	"net"
+	"net/url"
 	"sync"
+	"syscall"
 
 	"github.com/googleapis/gnostic/OpenAPIv2"
 
+	errorsutil "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/version"
 	"k8s.io/client-go/discovery"
 	restclient "k8s.io/client-go/rest"
 )
 
+type cacheEntry struct {
+	resourceList *metav1.APIResourceList
+	err          error
+}
+
 // memCacheClient can Invalidate() to stay up-to-date with discovery
 // information.
 //
-// TODO: Switch to a watch interface. Right now it will poll anytime
-// Invalidate() is called.
+// TODO: Switch to a watch interface. Right now it will poll after each
+// Invalidate() call.
 type memCacheClient struct {
 	delegate discovery.DiscoveryInterface
 
 	lock                   sync.RWMutex
-	groupToServerResources map[string]*metav1.APIResourceList
+	groupToServerResources map[string]*cacheEntry
 	groupList              *metav1.APIGroupList
 	cacheValid             bool
 }
 
 // Error Constants
 var (
-	ErrCacheEmpty    = errors.New("the cache has not been filled yet")
 	ErrCacheNotFound = errors.New("not found")
 )
 
 var _ discovery.CachedDiscoveryInterface = &memCacheClient{}
 
+// isTransientConnectionError checks whether given error is "Connection refused" or
+// "Connection reset" error which usually means that apiserver is temporarily
+// unavailable.
+func isTransientConnectionError(err error) bool {
+	urlError, ok := err.(*url.Error)
+	if !ok {
+		return false
+	}
+	opError, ok := urlError.Err.(*net.OpError)
+	if !ok {
+		return false
+	}
+	errno, ok := opError.Err.(syscall.Errno)
+	if !ok {
+		return false
+	}
+	return errno == syscall.ECONNREFUSED || errno == syscall.ECONNRESET
+}
+
+func isTransientError(err error) bool {
+	if isTransientConnectionError(err) {
+		return true
+	}
+
+	if t, ok := err.(errorsutil.APIStatus); ok && t.Status().Code >= 500 {
+		return true
+	}
+
+	return errorsutil.IsTooManyRequests(err)
+}
```

Comment:
If the group version doesn't exist, server will return 404, that's the only permanent error, so the client can retry any error other than 404. WDYT?

Reply:
I think it's better for this list to be incomplete rather than with too many elements: I trying to avoid a case when bug or misconfiguration creates retries calls at some high rate generating significant load on apiserver.

Re retrying any error != 404 -- I don't think this is a good idea to do that --  errors like 401 (unauthorized), 403, 422 (invalid request) shouldn't be retried in most of the cases.

Also I'm afraid that there can be some other errors: e.g. IIUC right now empty response can mean cache miss:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/discovery/discovery_client.go#L182

This response is changed in d.serverResourcesForGroupVersion to fmt.Errorf response. We definitely don't want to retry that.

Reply:
> I think it's better for this list to be incomplete rather than with too many elements

I agree with this.

Though I don't think we will get other errors in reality. Because a `groupVersion` needs to get into the cache in the first place to be retried, that means the apiserver has listed the `groupVersion` in its discovery doc.

The line you pasted is for compatibility with v1.0 k8s servers, hopefully there are not many of them in production.

Anyway, I agree that being defensive is right.